core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java

```diff
@@ -0,0 +1,245 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.RelFieldCollation;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.logical.LogicalWindow;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rel.type.RelRecordType;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+
+/**
+ * Planner rule that pushes
+ * a {@link org.apache.calcite.rel.logical.LogicalProject}
+ * past a {@link org.apache.calcite.rel.logical.LogicalWindow}.
+ */
+public class ProjectWindowTransposeRule extends RelOptRule {
+  /** The default instance of
+   * {@link org.apache.calcite.rel.rules.ProjectWindowTransposeRule}. */
+  public static final ProjectWindowTransposeRule INSTANCE = new ProjectWindowTransposeRule();
+
+  private ProjectWindowTransposeRule() {
+    super(
+        operand(
+            LogicalProject.class,
+                operand(LogicalWindow.class,
+                    any())));
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final LogicalProject project = call.rel(0);
+    final LogicalWindow window = call.rel(1);
+    final List<RelDataTypeField> rowTypeWindowInput = window.getInput().getRowType().getFieldList();
+    final int windowInputColumn = rowTypeWindowInput.size();
+
+    // Record the window input columns which are actually referred
+    // either in the LogicalProject above LogicalWindow or LogicalWindow itself
+    final BitSet beReferred = findReference(project, window);
+
+    // If all the the window input columns are referred,
+    // it is impossible to trim anyone of them out
+    if (beReferred.cardinality() == windowInputColumn) {
+      return;
+    }
+
+    // Put a DrillProjectRel below LogicalWindow
+    final List<RexNode> exps = new ArrayList<>();
+    final List<RelDataTypeField> fields = new ArrayList<>();
+
+    // Keep only the fields which are referred
+    for  (int index = beReferred.nextSetBit(0); index >= 0;
+        index = beReferred.nextSetBit(index + 1)) {
+      final RelDataTypeField relDataTypeField = rowTypeWindowInput.get(index);
+      exps.add(new RexInputRef(index, relDataTypeField.getType()));
+      fields.add(relDataTypeField);
+    }
+
+    final LogicalProject projectBelowWindow = new LogicalProject(
+        window.getCluster(),
+        window.getTraitSet(),
+        window.getInput(),
+        exps,
+        new RelRecordType(fields));
```

Comment:
Use below to create type:
final RelDataType rowtype = RexUtil.createStructType(
        window.getCluster().getTypeFactory(),
        exps,
        fieldNames);